crates/core/src/host/instance_env.rs

```diff
@@ -426,9 +426,115 @@ impl InstanceEnv {
             Code::Table(table) => table,
             _ => unreachable!("query should always return a table"),
         };
-        Ok(std::iter::once(bsatn::to_vec(&row_type))
-            .chain(results.data.into_iter().map(|row| bsatn::to_vec(&row)))
-            .map(|bytes| bytes.expect("encoding algebraic values should never fail")))
+        Ok(BSatnCompactor::new(
+            std::iter::once(Ok(IterItem::Header(row_type)))
+                .chain(results.data.into_iter().map(|row| Ok(IterItem::Row(row)))),
+        ))
+    }
+}
+
+/// ==== Iterator packing ====
+
+/// An item yielded by a table iterator, either a header describing the type of the table
+/// or a row with data. Iterators yielding this type should return exactly one header
+/// at the beginning of iteration.
+enum IterItem {
+    Header(ProductType),
+    Row(ProductValue),
+}
+
+/// State of a compactor iterator
+#[derive(Copy, Clone)]
+enum BSatnCompactorState {
+    Start,
```

Comment:
Short comments on these state machine states help a lot with. The names are good, so I think most of what you are trying to capture is what can transition to what. It then helps with reading the giant loop/switch, to know that e.g Complete isn't supposed to transition to anything else, so the unreachable!() is warranted etc.